File: all-submissions/2038-nearest-exit-from-entrance-in-maze/solution.java


import static java.lang.Math.max;

class Solution {
    class State{
        int row;
        int col;
        int steps;
        State(int row, int col,int  steps) {
            this.row = row;
            this.col = col;
            this.steps = steps;
        }
    }

    public int nearestExit(char[][] maze, int[] entrance) {

        Queue<State> cellsQueue = new LinkedList<>();
        State entranceState = new State(entrance[0], entrance[1], 0);
        cellsQueue.add(entranceState);
        int[][] directions = new int[][]{{0,1}, {1,0}, {0,-1}, {-1,0}};
        int rowLength = maze.length;
        int colLength = maze[0].length;
        boolean[][] seen = new boolean[rowLength][colLength];
        seen[entrance[0]][entrance[1]] = true;

        while(!cellsQueue.isEmpty()){
            State currState = cellsQueue.poll();
            int row = currState.row;
            int col = currState.col;
            int steps = currState.steps;
            for(int[] direction: directions) {
                int nextRow = row +direction[0];
                int nextCol = col + direction[1];
                if(isValid(nextRow, nextCol, rowLength, colLength, maze) && !seen[nextRow][nextCol]){
                    if(isExit(nextRow, nextCol, rowLength, colLength)) {
                        return steps+1;
                    }
                    seen[nextRow][nextCol] = true;
                    cellsQueue.add(new State(nextRow, nextCol, steps+1));
                }
            }
        }
        return -1;
    }

    boolean isValid(int row, int col, int rowLength, int colLength, char[][]maze){
        return (row >=0 && row<rowLength && col >=0 && col < colLength && maze[row][col] != '+');
    }

    boolean isExit(int row, int col, int rowLength, int colLength){
        return (row == 0 || row == rowLength-1 || col ==0 || col == colLength-1);
    }
}

File: all-submissions/2038-nearest-exit-from-entrance-in-maze/solution.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        queue = collections.deque()
        visited = set()
        queue.append((entrance[0], entrance[1], 0))
        visited.add((entrance[0], entrance[1]))
        ROWS = len(maze)
        COLS = len(maze[0])
        directions = [[0,1],[1,0], [-1,0], [0,-1]]

        def isValid(r,c):
            return 0<=r<ROWS and 0<=c<COLS and maze[r][c] == "."

        def isExit(r,c):
            return r ==0 or r== ROWS-1 or c == 0 or c== COLS-1

        while queue:
            row, col, steps = queue.popleft()
            
            for dr, dc in directions:
                nr = dr + row
                nc = dc + col
                if isValid(nr,nc) and (nr, nc)  not in visited:
                    if isExit(nr, nc):
                        return steps+1
                    queue.append((nr,nc, steps+1))
                    visited.add((nr,nc))
        return -1
